ios/uPortMobile/Classes/EthereumSigner.m

```diff
@@ -0,0 +1,203 @@
+    BTCMutableBigNumber* k = [[BTCMutableBigNumber alloc] initWithUnsignedBigEndian:kdata];
+    [k mod:n]; // make sure k belongs to [0, n - 1]
+    
+    if (BN_cmp(k.BIGNUM, [BTCBigNumber zero].BIGNUM) <= 0 && BN_cmp(k.BIGNUM, n.BIGNUM) >= 0) return NULL;
+
+    BTCDataClear(kdata);
+  
+    BTCCurvePoint* K = [[BTCCurvePoint generator] multiply:k];
+    // K was infinity and does not work, redo it
+    if ([K isInfinity])  return NULL;
+    BTCMutableBigNumber* Kx = [K.x mutableCopy];
+    [Kx mod: n];
+    
+    BTCBigNumber* hashBN = [[BTCBigNumber alloc] initWithUnsignedBigEndian:hash];
+  
+    // Compute s = (k^-1)*(h + Kx*privkey)
+  
+    BTCMutableBigNumber* signatureBN = [[[[privkeyBN mutableCopy] multiply:Kx mod:n] add:hashBN mod: n] multiply:[k inverseMod:n] mod:n];
+
+    // Neither r nor s can be zero
+    if ([Kx isZero] || [signatureBN isZero]) return NULL;
+    BTCBigNumber *r = [Kx copy];
+    uint8_t recoveryParam = BN_is_odd(K.y.BIGNUM) ? 1 : 0;
+    
+    [privkeyBN clear];
+    [k clear];
+    [hashBN clear];
+    [K clear];
+    [Kx clear];
+
+    BTCBigNumber *twiceS = [[signatureBN mutableCopy] add: signatureBN];
+    
+    BTCBigNumber *s;
+    if (lowS && BN_cmp(twiceS.BIGNUM, n.BIGNUM) > 0) {
+      // enforce low S values, by negating the value (modulo the order) if above order/2.
+      s = [[n mutableCopy] subtract: signatureBN]; // Maybe this should be swapped
+      recoveryParam ^= 1;
+    } else {
+      s = [signatureBN copy];
+    }
+
+    [twiceS clear];
+    [signatureBN clear];
+
+    NSData* rData = [r unsignedBigEndian];
+    NSData* sData = [s unsignedBigEndian];
+    [r clear];
+    [s clear];
+    
+    return @{
+             @"r": rData,
+             @"s": sData,
+             @"recoveryParam": @(recoveryParam)
+             };
+}
+
+NSData *simpleSignature(BTCKey *keypair, NSData *hash) {
+  NSDictionary *sig = genericSignature(keypair, hash, NO);
+  if (!sig) return NULL;
+  NSData *rData = (NSData *)sig[@"r"];
+  NSData *sData = (NSData *)sig[@"s"];
+  ///////
+  NSMutableData *sigData = [NSMutableData dataWithLength:64];
+  unsigned char* sigBytes = sigData.mutableBytes;
+  memset(sigBytes, 0, 64);
+  
+  memcpy(sigBytes, rData.bytes, 32);
+  memcpy(sigBytes+32, sData.bytes, 32);
+  return sigData;
+}
+
+// Not used here but leave in for use in library when pulling out again
+static int ECDSA_SIG_recover_key_GFp(EC_KEY *eckey, BIGNUM *r, BIGNUM *s, const unsigned char *msg, int msglen, int recid, int check) {
+  if (!eckey) return 0;
+  
+  int ret = 0;
+  BN_CTX *ctx = NULL;
+  
+  BIGNUM *x = NULL;
+  BIGNUM *e = NULL;
+  BIGNUM *order = NULL;
+  BIGNUM *sor = NULL;
+  BIGNUM *eor = NULL;
+  BIGNUM *field = NULL;
+  EC_POINT *R = NULL;
+  EC_POINT *O = NULL;
+  EC_POINT *Q = NULL;
+  BIGNUM *rr = NULL;
+  BIGNUM *zero = NULL;
+  int n = 0;
+  int i = recid / 2;
+  
+  const EC_GROUP *group = EC_KEY_get0_group(eckey);
+  if ((ctx = BN_CTX_new()) == NULL) { ret = -1; goto err; }
```

Comment:
It may help remove some visual clutter here to use a good old CPP macro:

```
#define CHECK(condition, error) { if (!(condition)) { ret = error; goto err; }

    CHECK(EC_GROUP_get_order(group, order, ctx)), -1)
    x = BN_CTX_get(ctx);
    CHECK(BN_copy(x, order)), -1)
    CHECK(BN_mul_word(x, i)), -1)
```